### PR TITLE
Wayland: added basic support for idle inhibit protocol

### PR DESCRIPTION
**Pull request by anriha**
This adds support for idle-inhibit-unstable-v1 protocol. It is dependant on flacjacket/pywlroots#75 
It works, but only for applications that only have one surface e.g. mpv for others like Firefox it does not. 

The problem is that the new idle inhibitor event only has reference to the surface it is bound to and currently I can not find any way to associate it with subsurfaces. For that it would require somehing like every window having references all the pointers of subsurfaces but that does not seem like good solution.
The other way would be to ignore if window is mapped and only care about if there is idle inhibitor. This would work but it wouldn't work as described in the protocol.

### CODE REVIEW

File: libqtile/backend/wayland/window.py
Comment (m-col) [inline libqtile/backend/wayland/window.py:302]: Could you rename `_listener` to `listener` and add a type annotation to it? The reasoning for "_listener" was because it is usually an unused variable that we want to ignore, whereas he it's being used.

File: libqtile/backend/wayland/core.py
Comment (m-col) [inline libqtile/backend/wayland/core.py:423]: You can iterate through all of a windows surfaces with a callback like this:

```suggestion
        for win in self.qtile.windows_map.values():
            if isinstance(win, window.Window):
		win.surface.for_each_surface(win.add_idle_inhibitor, idle_inhibitor)
		if idle_inhibitor.data:
		    break
```
In this approach, you also need to adjust `Window.add_idle_inhibitor`:
```py
    def add_idle_inhibitor(self, surface: surface.Surface, _x, _y, inhibitor: IdleInhibitorV1) -> None:
        if surface == inhibitor.surface:
            inhibitor.data = self
            self._idle_inhibitors.append(inhibitor)
            self.add_listener(inhibitor.destroy_event, self._on_inhibitor_destroy)
            if len(self._idle_inhibitors) == 1:
                self.core.check_idle_inhibitor()
```
Then we are using `inhibitor.data` to know if the corresponding surface and window has been found.

File: libqtile/backend/wayland/window.py
Comment (m-col) [inline libqtile/backend/wayland/window.py:82]: If we don't need the objects themselves, maybe we can get away with just using an `int` as a counter, and increment/decrement it?

File: libqtile/backend/wayland/window.py
Comment (m-col) [inline libqtile/backend/wayland/window.py:168]: Is this because of another class inheriting this method, which might not have this attribute? If so, this hack will fortunately not be necessary soon by having more well-defined types: https://github.com/qtile/qtile/pull/3293

File: libqtile/backend/wayland/core.py
Comment (m-col) [inline libqtile/backend/wayland/core.py:171]: ` _idle_ihibitor_manager` -> ` idle_ihibitor_manager` just to be consistent with other local-only variables

File: libqtile/backend/wayland/core.py
Comment (m-col) [inline libqtile/backend/wayland/core.py:804]: I think the generator for loop hurts readability:
```suggestion
        for win in self.mapped_windows:
	    if isinstance(win, window.Window) and win.is_idle_inhibited:
		self.idle.set_enabled(self.seat, True)
		return

	self.idle.set_enabled(self.seat, False)
```

**Review by m-col — CHANGES_REQUESTED** [review]
Thanks again for working on improving the Wayland backend. Overall this is looking great - I've left some review comments, which I think will also address your comment about not acccounting for subsurfaces.

File: libqtile/backend/wayland/core.py
Comment (anriha) [reply libqtile/backend/wayland/core.py]: Nice, that works!

File: libqtile/backend/wayland/window.py
Comment (anriha) [reply libqtile/backend/wayland/window.py]: Yep that is probably easier, commited.

File: libqtile/backend/wayland/window.py
Comment (anriha) [reply libqtile/backend/wayland/window.py]: Hmm that would solve this, it is present on several places in this PR, should I wait until your change is merged, or should I leave it as is and change it later?

Comment (m-col) [reply libqtile/backend/wayland/window.py]: I'm happy to merge this first and rebase that other PR on top of this. However, we do need to account for the different window classes, specifically we need to ensure that if `Window.cmd_static` is called to turn an XDG shell window into a `Static` window, that window's `_idle_inhibitors` also needs to be copied to the new object. Similarly we need to register idle inhibitors that are created for windows represented by `Static` and handle those just the same if the window is part of the layer shell or xdg shell. If it makes things easier to see how these things are represnted then waiting until that other PR is merged might make sense. Up to you.

Comment (anriha) [reply libqtile/backend/wayland/window.py]: Ok, I've added idle_inhibitors_count to other window classes, and to cmd_static. It seems to work fine.

**Comment by anriha** [comment]
Pywlroots was merged so this should now work.

File: libqtile/backend/wayland/window.py
Comment (m-col) [inline libqtile/backend/wayland/window.py:389]: How come the type of inhibitor is `IdleInhibitorV1 | None` and not just `IdleInhibitorV1`?

File: libqtile/backend/wayland/core.py
Comment (m-col) [inline libqtile/backend/wayland/core.py:798]: The type of `self.mapped_windows` is `window.WindowType` so I think this `isinstance` test is redundant.

File: libqtile/backend/wayland/window.py
Comment (anriha) [reply libqtile/backend/wayland/window.py:389]: Because the type signature for all implementations of for_each_surface is `data: T | None` so without that mypy complains.

Comment (m-col) [reply libqtile/backend/wayland/window.py:389]: OK cool, thanks

File: libqtile/backend/wayland/core.py
Comment (anriha) [reply libqtile/backend/wayland/core.py]: You are right that is redundant, fixed.

**Review by m-col — APPROVED** [review]

**Comment by m-col** [comment]
Thanks again for this! Auto merge enabled.

**Pull request merged**
